src/v/kafka/server/quota_manager.cc

```diff
@@ -36,55 +58,104 @@ quota_manager::quota_manager()
+
+    vassert(_client_quotas, "_client_quotas should have been initialized");
+
+    auto it = _client_quotas->find(qid);
+    auto inserted = false;
+    if (it == _client_quotas->end()) {
+        auto produce_rate = get_client_target_produce_tp_rate(qid);
+        auto fetch_rate = get_client_target_fetch_tp_rate(qid);
+        auto partition_mutation_rate = *_target_partition_mutation_quota();
+        auto default_num_windows = static_cast<uint32_t>(
+          _default_num_windows());
+        auto default_window_width = _default_window_width();
+        auto replenish_threshold = static_cast<uint64_t>(
+          _replenish_threshold().value_or(1));
+
+        auto update_func =
+          [produce_rate,
+           fetch_rate,
+           partition_mutation_rate,
+           default_num_windows,
+           default_window_width,
+           replenish_threshold,
+           qid,
+           now](const client_quotas_map_t& old_map) -> client_quotas_map_t {
+            auto new_map = client_quotas_map_t{old_map};
+
+            auto produce_tracker = make_bucket(
+              produce_rate, replenish_threshold);
+            auto fetch_tracker = fetch_rate.has_value() ? make_bucket(
+                                   *fetch_rate, replenish_threshold)
+                                                        : std::nullopt;
+
+            new_map.emplace(
+              qid,
+              client_quota{
+                now,
+                clock::duration(0),
+                produce_tracker,
+                fetch_tracker,
+                /// pm_rate is only accessed on the qm home shard
+                std::optional<token_bucket_rate_tracker>(
+                  {partition_mutation_rate,
+                   default_num_windows,
+                   default_window_width})});
+
+            return new_map;
+        };
+
+        co_await seastar::smp::submit_to(
+          _client_quotas.shard_id(),
+          [&update_func,
+           _client_quotas = std::ref(this->_client_quotas)]() -> ss::future<> {
+              co_await _client_quotas.get().update(update_func);
+              co_return;
+          });
+
+        inserted = true;
+        it = _client_quotas->find(qid);
+        if (it == _client_quotas->end()) {
+            // The newly inserted quota map entry should always be available
+            // here because the update to the map is guarded by a mutex, so
+            // there is no chance of losing updates. There is a low chance that
+            // we insert into the map, then there is an unusually long pause on
+            // the handler core, gc() gets scheduled and cleans up the inserted
+            // quota, and by the time the handler gets here the inserted entry
+            // is gone. In that case, we give up.
+            vlog(klog.debug, "Failed to find quota id after insert...");
+            throw std::runtime_error("Failed to find quota id after insert");
```

Comment:
Based on some local testing I've done, we close the connection in this case and the client has to reconnect.
```
DEBUG 2024-05-17 09:45:38,776 [shard 0:main] kafka - server.cc:310 - max_reauth_ms: 0 // <-- Kafka accepts the connection
DEBUG 2024-05-17 09:45:38,777 [shard 0:main] kafka - connection_context.cc:125 - Connection input_shutdown; aborting operations for 172.26.0.29:38986
DEBUG 2024-05-17 09:45:38,778 [shard 0:main] kafka - connection_context.cc:125 - Connection input_shutdown; aborting operations for 172.26.0.29:38992
WARN  2024-05-17 09:45:38,779 [shard 0:main] kafka - server.cc:407 - Error kafka rpc protocol 172.26.0.29:38992: std::runtime_error (Failed to find quota id after insert)
ERROR 2024-05-17 09:45:38,779 [shard 0:main] kafka - server.cc:159 - Error[applying protocol] remote address: 172.26.0.29:38992 - std::runtime_error (Failed to find quota id after insert)
DEBUG 2024-05-17 09:45:38,883 [shard 0:main] kafka - server.cc:310 - max_reauth_ms: 0 // <-- Kafka accepts the re-connection
```
I think that's fine because we really don't expect this to happen. The post-await code here needs to be delayed for the delay interval (~10s by default) and needs to be reordered with respect to the update done by `gc()`. So I'm not even sure it's possible because I would assume that the scheduler would not reorder this code wrt updates issued later.

But the alternatives I can think of are (a) retrying here more to reduce the chances of this happening further, or (b) handling this failure one stack up but still within `quota_manager` and just not applying any delays in this case. Both of these make the code more complex for the sake of this rare case, but I guess (b) would be the most reliable way of handling this (limiting failure to quotas). What do you think?

Reply:
I've changed this now so that we just don't throttle in this case.